source/MRViewer/MRSpaceMouseHandlerWindows.cpp

```diff
@@ -213,22 +217,61 @@ void SpaceMouseHandlerWindows::updateConnected_()
         {
             if ( buttons_[i] ) // button up
                 viewer.spaceMouseUp( mapButtons_[i] );
-        }
-        buttons_ = {};
-    }
+		}
+        std::for_each( buttons_.begin(), buttons_.end(), [](auto& v) { v = 0; } );
 
-    joystickIndex_ = newJoystickIndex_;
+        if ( newJoystickIndex == -1 && updateThread_.joinable() )
+            updateThread_.join();
+    }
 
-    if ( joystickIndex_ != -1 )
+    if ( newJoystickIndex != -1 )
     {
         int count;
-        const float* axesNew = glfwGetJoystickAxes( joystickIndex_, &count );
-        std::copy( axesNew, axesNew + 6, axes_.begin() );
+        const float* axesNew = glfwGetJoystickAxes( newJoystickIndex, &count );
+        std::copy( axesNew, axesNew + 6, axesOld_.begin() );
+
+        if ( joystickIndex_ == -1 )
+            startUpdateThread_();
     }
 
+	joystickIndex_ = newJoystickIndex;
     getViewerInstance().mouseController.setMouseScroll( joystickIndex_ == -1 );
 }
 
+void SpaceMouseHandlerWindows::startUpdateThread_()
+{
+    updateThreadActive_ = true;
+    axesDiff_ = { 0, 0, 0, 0, 0, 0 };
+	updateThread_ = std::thread( [&]
+    {
+        std::array<float, 6> axesDiff;
+        std::for_each( axesDiff.begin(), axesDiff.end(), [](float& v) { v = 0.f; } );
+       int count = 0;
+       do
+       {
+           if ( active_ && joystickIndex_ != -1 )
+		   {
+               const float* axesNew = glfwGetJoystickAxes( joystickIndex_, &count );
+               if ( count == 6 )
+			   {
+				   axesDiff = axesDiff_;
+                   for ( int i = 0; i < 6; ++i )
+                   {
+                       float newDiff = axesNew[i] - axesOld_[i];
+                       if ( newDiff * axesDiff[i] < 0 )
+                           axesDiff[i] = newDiff;
+                       else if ( std::fabs( axesDiff[i] ) < std::fabs( newDiff ) )
+                           axesDiff[i] = newDiff;
+                   }
+                   axesDiff_ = axesDiff;
```

Comment:
How do we check that main thread have read the previous value of `axesDiff_` before writing new one here? I thin it is better to compute the diff in main thread